### ISSUE TITLE
ICML 26 | Latent Collaboration in Multi-Agent Systems

### ISSUE THREAD

**Issue by Dszdsxc0405** [AI-Infra-Team/awesome-papers#219]
## 0 Paper Information
**Title:** Latent Collaboration in Multi-Agent Systems
**Authors:** iaru Zou 1 2 * † Ruizhong Qiu 3 † Gaotang Li 3 † Xiyuan Yang 3 † Katherine Tieu 3 † Pan Lu 2 † Ke Shen  Hanghang Tong 3 Yejin Choi 2 Jingrui He 3 B James Zou 2 B Mengdi Wang 1 B Ling Yang
**Institudes:** 1Princeton University；2Stanford University；3UIUC.
**Paper:** [[[2511.20639] Latent Collaboration in Multi-Agent Systems](https://arxiv.org/abs/2511.20639)](https://arxiv.org/abs/2511.20639)
**Github Note Link:** [[Gen-Verse/LatentMAS: [ICML 2026 Spotlight] Latent Collaboration in Multi-Agent Systems](https://github.com/Gen-Verse/LatentMAS)](https://github.com/Gen-Verse/LatentMAS)
**Openreview:** [[Latent Collaboration in Multi-Agent Systems | OpenReview](https://openreview.net/forum?id=syG9I9ofd8)](https://openreview.net/forum?id=syG9I9ofd8)
**Author Page:**  [[Jiaru "Rubin" Zou (@Jiaru_Zou) / X](https://x.com/Jiaru_Zou)](https://x.com/Jiaru_Zou)

## 1. Background and Problem

As large language models evolve from single-model reasoning toward **Multi-Agent Systems (MAS)**, complex tasks are increasingly completed through collaboration among multiple agents with different roles. For example, in a sequential MAS, a Planner develops a solution strategy, a Critic identifies potential errors, a Refiner revises the reasoning process, and a Solver produces the final answer. In a hierarchical MAS, domain experts in mathematics, science, and programming reason independently, after which a Summarizer aggregates their outputs.

<img width="397" height="282" alt="Image" src="https://github.com/user-attachments/assets/82da99fc-f1bf-4520-ab9c-03b8f7bf9ac9" />

Most existing multi-agent systems use natural language as the shared communication medium. A preceding agent first decodes its internal reasoning state into text. The next agent then reads this text and converts it back into internal representations through tokenization, embedding, and prefill computation. The overall process can be summarized as follows:

```text
Agent 1's internal representation
→ autoregressive token-by-token reasoning text generation
→ Agent 2 re-encodes the text
→ Agent 2 continues generating text
→ subsequent agents repeat the same process
```

Although text-based collaboration offers good interpretability and cross-model compatibility, it introduces two major problems.

The first is an **information representation bottleneck**. LLMs internally represent semantic information using high-dimensional continuous vectors, whereas text tokens are discrete symbols. Mapping a continuous hidden state to a single token effectively compresses a rich semantic representation into one discrete choice from a finite vocabulary. Some information may therefore be lost during this conversion.

The second is the high **reasoning and communication overhead**. Every agent must perform token-by-token autoregressive decoding, while subsequent agents must repeatedly process the long text generated by preceding agents. In mathematical and scientific reasoning tasks, a single agent may generate thousands or even tens of thousands of chain-of-thought tokens. As a result, the latency and computational cost of a multi-agent system can grow rapidly.

Previous studies have separately explored latent reasoning within a single model and communication through hidden states or KV caches. However, there has not been a complete multi-agent framework that supports both latent reasoning and latent communication. The paper therefore raises the following central question: **Can multiple LLM agents collaborate entirely in a continuous latent space without relying on intermediate text?**

---

## 2. Observations and Opportunities

The paper’s basic observation is that the information used by an LLM during reasoning primarily exists in the Transformer’s internal **hidden states**, while text is only an external representation obtained by projecting hidden states into the vocabulary space.

In standard text generation, the final-layer hidden state at the current position, denoted as (h_t), is passed through the output projection and Softmax to calculate token probabilities. After selecting a token, the model converts that token back into an embedding for the next generation step. This process can be expressed as:

$$  
h_t  
\rightarrow  
\text{vocabulary probabilities}  
\rightarrow  
\text{token}  
\rightarrow  
\text{embedding}  
\rightarrow  
\text{next step}  
$$

The authors argue that this “vector–text–vector” conversion is not necessary for internal model reasoning. In principle, the continuous hidden state can be used directly to drive the next computation step, thereby avoiding discretization and explicit text decoding.

The authors further examine whether latent representations contain meaningful semantics. On 300 MedQA questions, they compare the final-layer hidden vectors generated by LatentMAS with the embeddings of text tokens generated by TextMAS. The results show that the two types of representations occupy broadly similar regions in the semantic space, indicating that latent vectors can encode semantics comparable to those expressed through textual reasoning. At the same time, the latent vectors cover a wider region, suggesting greater representational diversity and potential expressive capacity.

<img width="1287" height="500" alt="Image" src="https://github.com/user-attachments/assets/c2d8ce82-8e8a-4828-bf29-9b6919d012bb" />

A second important observation is that directly feeding final-layer hidden states back into the model causes an **input–output distribution mismatch**. Final-layer hidden states belong to a deep semantic representation space, whereas token embeddings belong to the shallow input space expected by the model. The two distributions are statistically different. Figure 6 shows that unprocessed hidden states deviate substantially from the normal input embedding region. If such vectors are repeatedly reused, representation drift can accumulate and eventually reduce reasoning quality.

<img width="1350" height="463" alt="Image" src="https://github.com/user-attachments/assets/d3642538-461a-4448-a342-2d988a3ee4ac" />

<img width="429" height="365" alt="Image" src="https://github.com/user-attachments/assets/f7dc43b1-f6e3-410f-a78c-84fac9f0e287" />

The third opportunity comes from the **KV cache**. When each latent step passes through all Transformer layers, it produces a corresponding Key and Value representation at every layer. Therefore, the KV cache can store not only the original question and role prompt, but also the newly generated latent thoughts. A subsequent agent can directly access these KVs through the attention mechanism, without converting the latent thoughts into text and performing another prefill pass.

---

## 3. Innovations and Solution

The paper proposes a completely training-free latent multi-agent framework called **LatentMAS**, which consists of three main mechanisms.

<img width="1001" height="707" alt="Image" src="https://github.com/user-attachments/assets/ebd0ac65-fb0d-481c-bef4-82053448017a" />

First, the paper introduces **Auto-regressive Latent Thoughts Generation**. At each latent step, an agent takes the final-layer hidden state at the current position, (h_t). Instead of decoding it into a token, the system applies an input–output alignment matrix (W_a):

$$  
e_{t+1}=h_tW_a  
$$

The aligned vector (e_{t+1}) is then used as the input representation at the next sequence position. After passing through all Transformer layers, the model produces a new hidden state (h_{t+1}). Repeating this procedure for (m) steps produces:

$$  
H=  
[h_{t+1},h_{t+2},\ldots,h_{t+m}]  
$$

This sequence of continuous vectors is defined as the agent’s **latent thoughts**. The alignment matrix is derived from the model’s existing input embedding matrix and output projection matrix. It is computed only once and does not require additional training. Experiments show that the alignment operation improves accuracy by 2.3 to 5.3 percentage points.

Second, the paper proposes **Latent Working Memory Transfer**. After an agent completes all latent steps, the system extracts the KV cache from every Transformer layer. This cache contains both the input context and the newly generated latent thoughts. The next agent concatenates the preceding agent’s KV cache with its own KV cache at the corresponding layers and directly reads the preceding agent’s internal information through attention:

<img width="383" height="72" alt="Image" src="https://github.com/user-attachments/assets/e1811d1a-666e-47bc-8974-50eb638db377" />

Therefore, Agent (A_2) does not need to reconstruct Agent (A_1)’s reasoning as text, nor does it need to recompute (A_1)’s context. Instead, it treats (A_1)’s KV cache as an already processed historical working memory. After (A_2) finishes reasoning, it passes an updated KV cache containing information from both (A_1) and (A_2) to (A_3).

Third, the framework requires **only the final agent to decode a textual answer**. All intermediate agents reason and communicate in latent space, thereby avoiding the generation of large amounts of intermediate chain-of-thought text. Compared with methods that optimize only latent reasoning or only KV-based communication, LatentMAS is the first framework to unify both components into an end-to-end multi-agent collaboration process. It is also applicable to different MAS architectures, including sequential and hierarchical systems.

---

## 4. Experiments and Results

The paper evaluates LatentMAS on nine benchmarks covering mathematical and scientific reasoning, commonsense reasoning, and code generation. These benchmarks include GSM8K, AIME24, AIME25, GPQA-Diamond, MedQA, ARC-Easy, ARC-Challenge, MBPP+, and HumanEval+. The evaluated backbone models include Qwen3-4B, Qwen3-8B, Qwen3-14B, as well as 3B- and 8B-scale Llama models. The baselines include a single agent, Sequential TextMAS, and Hierarchical TextMAS. Experiments are conducted on eight NVIDIA A100-80GB GPUs, and each agent performs 0, 10, 20, 40, or 80 latent steps.

<img width="1281" height="671" alt="Image" src="https://github.com/user-attachments/assets/3b03df2d-3911-4ac0-99e6-684458b49cc9" />

The overall results show that, compared with a single model, LatentMAS improves average accuracy by 14.6% under the sequential setting and by 13.3% under the hierarchical setting. Compared with TextMAS using the same architecture, the average accuracy improves by 2.8% and 4.6%, respectively.

In terms of efficiency, LatentMAS achieves an average speedup of approximately (4\times) over Sequential TextMAS and approximately (4.3\times) over Hierarchical TextMAS. It also reduces system-wide output token usage by 70.8% and 83.7%, respectively.

<img width="1291" height="425" alt="Image" src="https://github.com/user-attachments/assets/890aaaba-e4c3-4406-8f45-63c3c5dd7f91" />

On difficult reasoning tasks, TextMAS may require more than 20,000 output tokens to complete a full chain-of-thought trajectory. In contrast, LatentMAS generally achieves comparable or higher accuracy using fewer than 50 latent steps. For example, on GPQA-Diamond under the hierarchical setting, LatentMAS achieves up to approximately (7\times) inference speedup while reducing token usage by more than 80%.

The paper also finds that increasing the number of latent steps does not always improve performance. Accuracy generally reaches its highest level at around 40 to 80 latent steps. When the number of steps is further increased to 160, performance may plateau or decline. This result indicates that excessive latent reasoning can also introduce redundant information and accumulated errors.

<img width="425" height="298" alt="Image" src="https://github.com/user-attachments/assets/e46c7385-c5d4-402e-9a20-9c2419dd0bfb" />

Overall, LatentMAS demonstrates that natural language is not the only possible communication medium among LLM agents. By combining continuous hidden-state reasoning with layer-wise KV-cache communication, the system can substantially reduce intermediate token generation and end-to-end latency while maintaining or even improving task accuracy. However, as the number of agents increases, the accumulated KV cache also increases, resulting in higher memory consumption and attention computation costs. Cross-model collaboration, cross-node communication, and large-scale agent expansion therefore remain important directions for future research.
